### PR TITLE
add option to use synthetic input data

### PR DESCRIPTION
**Pull request by alfuyao-amd**

### CODE REVIEW

File: torchtitan/datasets/hf_datasets.py [hunk 1]
```diff
@@ -7,20 +7,53 @@
...
+    ):
+        self.batch_size = batch_size
+        self.seq_len = seq_len
+        self.vocab_size = vocab_size
+        self.device = device
+        self.dtype = dtype
+
+    def __iter__(self) -> Iterator[Tuple[Dict[str, torch.Tensor], torch.Tensor]]:
+        while True:
+            inputs = torch.randint(
```
Comment (tianyu-l) [inline torchtitan/datasets/hf_datasets.py:41]: This is fake data, not "synthetic" data.

**Review by tianyu-l — CHANGES_REQUESTED** [review]
Please justify the value of this change, following https://github.com/pytorch/torchtitan/blob/main/CONTRIBUTING.md#proof-of-value

In particular, why fake data is better than the default `c4` / `c4_test`?

File: torchtitan/datasets/hf_datasets.py (hunk 1 above)
Comment (alfuyao-amd) [reply torchtitan/datasets/hf_datasets.py:41]: Oh, how about call it random data? The goal is to remove dataset dependency for quick performance benchmarking.

Comment (alfuyao-amd) [reply torchtitan/datasets/hf_datasets.py:41]: Default c4 will have two problems:
1. Although small enough, but still have dependency for user to download before testing, and in rapid debugging and reruns, it is possible to hit HF request limit. Other case is in an unstable network, also affecting smooth development. I had to make local changes like this so I can developing without worry about dataset. Guess many users also has similar experience. 
2. With larger models and bigger batch size runs, it will easily loopback data, but same reason as 1, it may be limited or time consuming in many cases for user to download very large dataset.

Comment (fegin) [reply torchtitan/datasets/hf_datasets.py:41]: Having a random dataset sounds reasonable but I would suggest to put it in a separate module/file. To make this random dataset more useful, it would be good to have an option to fix the random seed so that users can use this random dataset to debug numerical issues, though this requires some efforts to ensure the random data won't explode the loss.

Comment (fegin) [reply torchtitan/datasets/hf_datasets.py:41]: Random dataset is usually very useful when debugging the CPU overhead brought by data loading, though I'm not sure if we already have such a use case. Multimodal may be benefit from random dataset.

Comment (tianyu-l) [reply torchtitan/datasets/hf_datasets.py:41]: @alfuyao1986 
> Although small enough, but still have dependency for user to download before testing, and in rapid debugging and reruns, it is possible to hit HF request limit. Other case is in an unstable network, also affecting smooth development. I had to make local changes like this so I can developing without worry about dataset. Guess many users also has similar experience.

We have `c4_test` stored in the repo
https://github.com/pytorch/torchtitan/tree/main/tests/assets/c4_test

> With larger models and bigger batch size runs, it will easily loopback data, but same reason as 1, it may be limited or time consuming in many cases for user to download very large dataset.

What would be the advantage of using random / fake data versus looping back on `c4_test`?

@fegin 
> Multimodal may be benefit from random dataset.

As we don't have multimodal training, I think the main thing I'd like to understand what's the benefit of adding random data on top of existing `c4_test`.

Comment (fegin) [reply torchtitan/datasets/hf_datasets.py:41]: Random dataset generally can skip the overhead of data loading, like actually reading from a disk. This is not related to whether the dataset is large or small. But as mentioned above, this may be more useful when we start to see dataloader overhead is a big thing. As for development efficiency, I didn't encounter such an issue, so I should not be the one to answer.

This is solely my opinion.

Comment (alfuyao-amd) [reply torchtitan/datasets/hf_datasets.py:41]: Oh, given c4_test is already pre-stored in the repo, for most of the cases, it should be fine. I am actually completely fine with using pre-stored c4_test dataset. Only two more consideration just bring up for discussion.
1. Random dataset can usually stress the whole stack better, numerically and computationally, vs. a small repeated dataset, but it is debatable that whether this additional stress practically realistic and necessary. 
2. Other frameworks (MaxText, Megatron-LM) do provide "synthetic/mock" data options for fast benchmarking, for ease of comparison point of view, it may be better to have a matching option.

Comment (tianyu-l) [reply torchtitan/datasets/hf_datasets.py:41]: Got it, thanks for the context!

From my perspective, the value of this dataset is somewhat limited, given we already have c4_test which doesn't involve randomness so has become a standard way for numerical testing even when parallelism / world size changes.

That said, if people have strong opinion to add this dataset, I'm OK, too. If that's the case, I would suggest making a new builder function & file, instead of piggyback on existing `build_hf_dataloader`. I understand that would make it harder to switch to this new dataset from config, but that's not a good reason to reuse.

Comment (fegin) [reply torchtitan/datasets/hf_datasets.py:41]: yes, we definitely shouldn't use `build_hf_dataloader` for random dataset. There is actually another benefit of random dataset (when it has a deterministic option) -- debugging checkpoint issue. Given that the dataloader is controlled by other package, having a random dataset with a deterministic option will make debugging checkpoint inconsistency easier, at least we can rule out the dataset/dataloader problem.
